larq/quantizers_test.py

```diff
@@ -325,3 +325,58 @@ def ste_grad(x):
             activation = lq.quantizers.DoReFaQuantizer(2)(tf_x)
         grad = tape.gradient(activation, tf_x)
         np.testing.assert_allclose(grad.numpy(), ste_grad(x))
+
+
+@pytest.mark.parametrize(
+    "quantizer",
+    [
+        ("ste_sign", lq.quantizers.SteSign),
+        ("approx_sign", lq.quantizers.ApproxSign),
+        ("ste_heaviside", lq.quantizers.SteHeaviside),
+        ("swish_sign", lq.quantizers.SwishSign),
+        ("magnitude_aware_sign", lq.quantizers.MagnitudeAwareSign),
+        ("ste_tern", lq.quantizers.SteTern),
+        ("dorefa_quantizer", lq.quantizers.DoReFaQuantizer),
+    ],
+)
+def test_metrics(quantizer):
+    quantizer_str, quantizer_cls = quantizer
+
+    # No metric
+    model = tf.keras.models.Sequential(
+        [lq.layers.QuantDense(3, kernel_quantizer=quantizer_str, input_shape=(32,))]
+    )
+    model.compile(loss="mse", optimizer="sgd")
+    assert len(model.layers[0]._metrics) == 0
+
+    # Metric added using scope
+    with lq.metrics.scope(["flip_ratio"]):
+        model = tf.keras.models.Sequential(
+            [lq.layers.QuantDense(3, kernel_quantizer=quantizer_str, input_shape=(32,))]
+        )
+    model.compile(loss="mse", optimizer="sgd")
```

Comment:
I'm trying to write a test:

```python
def test_flip_ratio_computation(quantizer):
    """Check whether weights that got flipped were correctly counted by FlipRatio."""
    quantizer_str, quantizer_cls = ("ste_sign", lq.quantizers.SteSign)

    # Model and data setup
    input_data = testing_utils.random_input((1, 100))
    with lq.metrics.scope(["flip_ratio"]):
        model = tf.keras.models.Sequential(
            [
                lq.layers.QuantDense(
                    1,
                    kernel_quantizer=quantizer_str,
                    input_shape=(100,),
                    use_bias=False,
                )
            ]
        )
    model.compile(loss="mse", optimizer="sgd")

    with lq.quantized_scope.scope(True):
        old_weights = np.array(model.layers[0].get_weights())

    model.fit(input_data, np.ones((1,)), epochs=1, batch_size=1)

    with lq.quantized_scope.scope(True):
        new_weights = np.array(model.layers[0].get_weights())

    expected = 1 - np.count_nonzero(old_weights == new_weights) / 100
    actual = float(model.layers[0].kernel_quantizer.metrics[0].result())
    assert np.allclose(expected, actual)
```

But this gets wildly wrong answers (e.g. 0.52 expected vs 0.81 actual). @MariaHeuss and I can't see where the test goes wrong, do you see something @lgeiger?

Reply:
Does it work if you cast the weights into integers before comparing them with `==` `np.count_nonzero(old_weights == new_weights)`?

Reply:
Nope, casting to integers by setting `old_weights` and `new_weights` with `np.array(model.layers[0].get_weights(), dtype=int)` doesn't change anything.

Reply:
Hmm, what if you use [`model.train_on_batch`](https://www.tensorflow.org/api_docs/python/tf/keras/Model?hl=en#train_on_batch) just to be sure?

Reply:
With this `train_on_batch(...)`, `actual` becomes `0.0`, and with `train_on_batch(..., reset_metrics=False)`, it's sometimes `0.0` and sometimes another seemingly random (incorrect) fraction.

Reply:
Update (cc @lgeiger): this test always yields `actual == 0.0` on `master`. I've tried tweaking it to run two batches, but that results in incorrect flip ratios; I've also tried running a batch right after `compile()` to warm it up and then doing the same, but that also results in incorrect flip ratios.

It seems the flip ratio metric behaves a bit unpredictably at the start of training, so it may be hard to get this test working. Should we drop the test or look into this in more detail later this week?

Reply:
Thanks for investigating. That might be the case since it not initialized correctly. Let's take another look at it next week.

Reply:
3a51945569f8a9f631f2abb00ac399eb4318d2bd should fix this.